fix(build): ensure amd bundles request `require` to be injected

dschmidt opened the pull request:
<!----------------------------------------------------------------------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md.
- Check that there isn't already a PR that solves the problem the same way. If you find a duplicate, please help us reviewing it.
- Update the corresponding documentation if needed.
- Include relevant tests that fail without this PR but pass with it.

Thank you for contributing to Vite!
----------------------------------------------------------------------->
<!-- What is this PR solving? Write a clear description or reference the issues it solves (e.g. `fixes #123`). What other alternatives have you explored? Are there any parts you think require more attention from reviewers? -->

### Description

make relative imports in amd/cjs relative to the current module instead of the baseURI.



Fixes #20860 

~~I've added an override to `customRelativeUrlMechanisms` to avoid modifying the copied code from rollup.~~
~~As far as I can tell rollup itself uses `module.uri` to replace `import.meta.url` in `amd` modules. So I assume it's okay to use it.~~
~~https://github.com/rollup/rollup/blob/ce6cb93098850a46fa242e37b74a919e99a5de28/src/ast/nodes/MetaProperty.ts#L209~~

~~An alternative could be to use `resolve.toUrl(module.id)` but it seems unneccessary if other parts of the code already rely on `module.uri`.~~

~~FWIW I've verified that `import.meta.url` gets transpiled to `new URL(module.uri, document.baseURI).href` in my setup with amd modules. Using `import.meta.url` directly for the imports here does not work as it happens too late in the transpilation process (getting errors that `module` may not be used outside of module scope).~~

~~I'm a bit lost regarding a regression test as I couldn't find a test for the relative path behavior that I could simply adjust... if you require me to implement one, could you give me a few pointers how to implement this specific one?~~

After creating a proper reproducer and discussions with @lukastaegert of rollup fame and @sapphi-red it turned out, joining the path with the baseURI is actually correct.
It works in rollup because modules request `require` to be injected. This local `require` instance has the correct base path and so `toUrl` returns a correct absolute path.
In vite this was missing, so we were falling back to the global `require`, so `toUrl` used baseUri as base instead of the current module's path.

This PR adds `require` to the dependencies of amd modules.


Unfortunately this change is breaking for a specific edge case: when a module already contains a local variable `require` the added parameter leads to an error. 
The `completeSystemJsWrapPlugin` has a similar problem.

dschmidt committed:
fix(build): make relative imports in amd/cjs relative to the current module instead of the baseURI

dschmidt commented:
I'm not sure about the baseUrl - is it possible to have multiple instances of requirejs?
We have multiple apps with individual baseUrls so to speak. 
If I do 
```
requirejs.config({
    baseUrl: '/assets/apps/unzip/'
  });
```

that's only valid for a single app but not for all others.

No idea why rollup implemented it this way. Maybe `module.uri` wasn't always available. 
Usually it probably just doesn't matter. I have to admit my use case is a bit of an edge case - but it's one that is supported by `system` and `es` modules.
I think it's a valid way to define "relative paths" to be relative to some base uri, but then all module systems should implement it that way.

dschmidt commented:
If you insist, I can also send a PR to Rollup and we can check their feedback 😅

sapphi-red commented:
Thank you for the examples and the PR to Rollup. I think it makes sense. Just in case, I'd like to wait for a week for the feedback on Rollup side.

dschmidt committed:
fix: ensure amd bundles request `require` to be injected

... to avoid falling back to the global require, which can break
relative import urls.

sapphi-red commented on packages/vite/src/node/plugins/completeAmdWrap.ts:
```diff
@@ -0,0 +1,30 @@
+
+      return {
+        code: code.replace(AmdWrapRE, (_, deps, params) => {
+          if (deps && /"require"/.test(deps)) {
```
```suggestion
          if (deps?.includes('"require"')) {
```

sapphi-red reviewed (commented):
Thank you. The plugin implementation looks good to me.

I've added some minor comments. Also would you add a comment here like the one for system format?
https://github.com/vitejs/vite/blob/2443e056941d63ea5988293ca1a1351af017e59d/packages/vite/src/node/build.ts#L1315-L1320
https://github.com/vitejs/vite/blob/2443e056941d63ea5988293ca1a1351af017e59d/packages/vite/src/node/build.ts#L1330

dschmidt committed:
Update packages/vite/src/node/plugins/completeAmdWrap.ts

Co-authored-by: 翠 <green@sapphi.red>

dschmidt committed:
Update packages/vite/src/node/plugins/completeAmdWrap.ts

Co-authored-by: 翠 <green@sapphi.red>

dschmidt committed:
fix: detect reject in deps when it's wrapped in single quotes

dschmidt committed:
test(amd): add amd playground

sapphi-red committed:
test: add assertions

sapphi-red committed:
Merge branch 'main' into fix/amd-relative-imports

sapphi-red committed:
chore: fix lockfile

sapphi-red reviewed (dismissed):
Thanks, I finished up the playground test.

sapphi-red committed:
chore: add more package.json fields

Co-authored-by: Bjorn Lu <bjornlu.dev@gmail.com>

sapphi-red commented on playground/amd/public/index.html:
It is because this file is not part of the build (the entrypoint is `index.ts`) and won't be copied to `dist` if it's in the root.

bluwy replied:
Alright, but it feels a bit weird to be building a Vite app this way. Usually the build output format doesn't affect dev, but I suppose it's fine as a test setup for now.

dschmidt commented:
Thanks everyone for the input and especially to @sapphi-red for taking the PR over the finish line :)
